File: backend/src/main/java/com/apocalypse/thefall/exception/GlobalExceptionHandler.java
package com.apocalypse.thefall.exception;

import jakarta.validation.ConstraintViolationException;
import org.springframework.context.MessageSource;
import org.springframework.context.i18n.LocaleContextHolder;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@RestControllerAdvice
public class GlobalExceptionHandler {
    private final MessageSource messageSource;

    public GlobalExceptionHandler(MessageSource messageSource) {
        this.messageSource = messageSource;
    }

    @ExceptionHandler(GameException.class)
    public ErrorResponse handleGameException(GameException ex) {
        String message = messageSource.getMessage(ex.getMessageKey(), ex.getArgs(), LocaleContextHolder.getLocale());
        return new ErrorResponse(ex.getStatus(), message);
    }

    @ExceptionHandler({ConstraintViolationException.class, IllegalArgumentException.class})
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ErrorResponse handleValidationException(Exception ex) {
        String message = messageSource.getMessage("error.validation", new String[]{ex.getMessage()},
                LocaleContextHolder.getLocale());
        return new ErrorResponse(HttpStatus.BAD_REQUEST, message);
    }

    @ExceptionHandler(Exception.class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public ErrorResponse handleGlobalException(Exception ex) {
        String message = messageSource.getMessage("error.unexpected", new String[]{ex.getMessage()},
                LocaleContextHolder.getLocale());
        return new ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, message);
    }
}

File: backend/src/main/java/com/apocalypse/thefall/exception/GlobalExceptionHandler.kt
package com.apocalypse.thefall.exception

import jakarta.validation.ConstraintViolationException
import org.springframework.context.MessageSource
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler(
    private val messageSource: MessageSource
) {

    @ExceptionHandler(GameException::class)
    fun handleGameException(ex: GameException): ErrorResponse {
        val message = messageSource.getMessage(
            ex.messageKey,
            ex.args,
            LocaleContextHolder.getLocale()
        )
        return ErrorResponse(ex.status, message)
    }

    @ExceptionHandler(ConstraintViolationException::class, IllegalArgumentException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleValidationException(ex: Exception): ErrorResponse {
        val message = messageSource.getMessage(
            "error.validation",
            arrayOf(ex.message ?: ""),
            LocaleContextHolder.getLocale()
        )
        return ErrorResponse(HttpStatus.BAD_REQUEST, message)
    }

    @ExceptionHandler(Exception::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    fun handleGlobalException(ex: Exception): ErrorResponse {
        val message = messageSource.getMessage(
            "error.unexpected",
            arrayOf(ex.message ?: ""),
            LocaleContextHolder.getLocale()
        )
        return ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, message)
    }
}
